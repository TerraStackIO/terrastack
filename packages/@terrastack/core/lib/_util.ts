export function removeEmpty(obj: any): any {
  if (obj == null) {
    return undefined;
  }

  if (typeof(obj) !== 'object') {
    return obj;
  }
  
  if (Array.isArray(obj)) {
    if (obj.length === 0) {
      return undefined;
    }

    return obj.map(x => {
      // do not remove "{}" and "[]" if they are array elements.
      if (Array.isArray(x) && x.length === 0) { return x; }
      if (x != null && typeof(x) === 'object' && Object.keys(x).length === 0) { return x; }
      return removeEmpty(x)
    });
  }

  if (obj.constructor.name !== 'Object') {
    throw new Error(`can't render non-simple object of type '${obj.constructor.name}'`);
  }

  if (Object.keys(obj).length === 0) {
    return undefined;
  }

  const newObj: { [key: string]: any } = { };

  for (const [key, value] of Object.entries(obj)) {
    const newValue = removeEmpty(value);
    if (newValue != null) {
      newObj[key] = newValue;
    }
  }

  return newObj;
}

export function snakeCase(str: string): string {
  if (!str) return '';

  return String(str)
    .replace(/^[^A-Za-z0-9]*|[^A-Za-z0-9]*$/g, '')
    .replace(/([a-z])([A-Z])/g, (_m, a, b) => a + '_' + b.toLowerCase())
    .replace(/[^A-Za-z0-9]+|_+/g, '_')
    .toLowerCase();
}