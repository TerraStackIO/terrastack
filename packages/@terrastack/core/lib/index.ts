export * from './resource-object';
export * from './app';
export * from './stack';
export * from './tf-reference';
