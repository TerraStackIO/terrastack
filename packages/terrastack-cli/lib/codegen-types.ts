import { JSONSchema4 } from "json-schema";
import { CodeMaker } from "codemaker";
import {  
  camelCase,
  pascalCase
} from "change-case";

export interface ResolvedTypes {
  type: string;
  assignable: boolean;
  optional: boolean;
  referencable: boolean;
}

function generatePropertyName(property: string): string {
  // `self` doesn't work in as property name in Python
  if (property === 'self') {    
    return camelCase(`${property}Property`)
  }
  // jsii can't handle `getFoo` properties, since it's incompatible with Java
  if (property.split('_')[0] === 'get') {    
    return camelCase(property.replace('get', 'fetch'))
  }
  return camelCase(property)
}


export class TypeGenerator {
  private readonly emitted = new Set<string>();
  public readonly types: {[key: string]: {[key: string]: ResolvedTypes}} = { };

  constructor(private readonly type: string) {
  }

  public addType(typeName: string, def: JSONSchema4) {    
    if (this.emitted.has(typeName)) {
      console.log('hier')
      return;
    }
    this.resolveTypes(typeName, def.block || def);    
  }

  public generate(code: CodeMaker) {
    for (const type of Object.keys(this.types)) {
      const spec = this.types[type];
      this.emitStruct(code, type, spec)
      code.line();
      this.emitted.add(type);
    }    
  }

  private resolveTypes(typeName: string, def: JSONSchema4) {
    const self = this;
    if (def.attributes || def.block_types) {
      for (const [ propName, propSpec ] of Object.entries(def.attributes || {})) {
        resolveProperty(propName, propSpec as JSONSchema4);
      }

      for (const [ blockName, blockSpec ] of Object.entries(def.block_types || {})) {
        const newTypeName = `${typeName}${pascalCase(blockName)}Props`
        self.addType(newTypeName, blockSpec as JSONSchema4)
        const emptyType = (Object.keys((blockSpec as JSONSchema4).block.attributes || {}).length === 0)
        addProperty(blockName, emptyType ? 'any' : newTypeName, true, true, false)
      }
    } else {
      for (const [ propName, propSpec ] of Object.entries(def || {})) {
        resolveProperty(propName, propSpec as JSONSchema4);
      }
    }

    function resolveProperty(name: string, def: JSONSchema4) {
      let assignable = true      
      if ((def.computed && !def.optional) || (self.type !== 'data' && name === 'id')) {
        assignable = false
      }
      const propertyType = self.typeForProperty(def, name, typeName);
      const optional = (def.optional || def.computed)
      addProperty(name, propertyType, assignable, optional, def.computed)
    }  

    function addProperty(name: string, propertyType: string, assignable: boolean, optional: boolean, referencable: boolean) {
      if (!self.types[typeName]) {
        self.types[typeName] = {}
      }

      self.types[typeName][name] = {
        type: propertyType,
        assignable,
        optional,
        referencable
      }
    }
  }

  private emitStruct(code: CodeMaker, typeName: string, def: {[key: string]: ResolvedTypes}) {
    code.openBlock(`export interface ${typeName}`);
    for (const propertyName of Object.keys(def)) {
      const property = def[propertyName]
      if (!property.assignable) continue;
      code.line(`readonly ${generatePropertyName(propertyName)}${property.optional ? '?' : ''}: ${property.type};`);
      code.line();
    }
    
    code.closeBlock();
  }

  private typeForProperty(def: JSONSchema4, name: string, typeName: string): string {  
    const comparable = def.type || def
    switch (true) {
      case (comparable === "string"): return 'string';
      case (comparable === "number"): return 'number';
      case (comparable === "integer"): return 'number';
      case (comparable?.toString() === "bool"): return 'boolean';
      case (Array.isArray(comparable)): return this.handleArrayType(comparable as JSONSchema4, name, typeName);
      default: 
        console.log({typeForProperty: def, name, typeName, comparable})
        return 'any';
    }
  }

  private handleArrayType(def: JSONSchema4, name: string, typeName: string): string {
    const element = ((def.type || def) as Array<any>)
    if (Array.isArray(element[element.length - 1])) {
      const type = element[0]
      const obj = element[element.length - 1]
      let newTypeName = undefined;
      if (this.jsonEqual(obj, ["map", "string"])) {
        newTypeName = `{[key: string]: string}`
      } else {
        newTypeName = `${typeName}${pascalCase(name)}Props`  
        this.addType(newTypeName, obj[obj.length - 1] as JSONSchema4)  
      }
      switch(type) {
        case 'list': return `${newTypeName}[]`;
        case 'set': return `${newTypeName}[]`;
      }    
    }
    switch (true) {
      case (this.jsonEqual(element, ["map", "string"])): return '{[k:string]: string}';
      case (this.jsonEqual(element, ["map", "bool"])): return '{[k:string]: boolean}';
      case (this.jsonEqual(element, ["map", "number"])): return '{[k:string]: number}';
      case (this.jsonEqual(element, ["set", "number"])): return 'number[]';
      case (this.jsonEqual(element, ["set", "string"])): return 'string[]';
      case (this.jsonEqual(element, ["list", "string"])): return 'string[]';      
      default: 
        console.log({handleArrayType: def, name, typeName})
        return 'any';
    }
  }

  private jsonEqual(a: any, b: any): boolean {
    return JSON.stringify(a) === JSON.stringify(b);
  }
}