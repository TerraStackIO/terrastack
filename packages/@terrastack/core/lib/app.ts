import { Construct, ConstructNode } from '@aws-cdk/core';
import * as fs from 'fs';
import * as path from 'path';

export interface AppOptions {
  /**
   * The directory to output Terraform files.
   * 
   * @default "dist"
   */
  readonly outdir?: string;
}

/**
 * Represents a Terrastack application.
 */
export class App extends Construct {
  /**
   * The output directory into which manifests will be synthesized.
   */
  public readonly outdir: string;

  /**
   * Defines an app
   * @param options configuration options
   */
  constructor(options: AppOptions = { }) {
    super(undefined as any, '');
    this.outdir = options.outdir ?? 'dist';
  }

  /**
   * Synthesizes all manifests to the output directory
   */
  public synth(): void {
    ConstructNode.synth(this.node, {
      outdir: this.outdir
    });

    // remove manifest.json and cdk.out
    rm(path.join(this.outdir, 'manifest.json'));
    rm(path.join(this.outdir, 'cdk.out'));
  }
}

function rm(filePath: string) {
  if (fs.existsSync(filePath)) {
    fs.unlinkSync(filePath);
  }
}