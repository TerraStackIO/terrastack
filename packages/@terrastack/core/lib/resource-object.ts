import { Construct, TagManager } from '@aws-cdk/core';
import {    
  snakeCase
} from "./_util";


/**
 * Metadata associated with this object.
 */
export interface ResourceObjectMetadata {
  /**
   * The unique, namespace-global, name of this object inside the Terraform
   * stack.
   *
   * Normally, you shouldn't specify names for objects and let the CDK generate
   * a name for you that is application-unique. The names CDK generates are
   * composed from the construct path components, separated by dots and a suffix
   * that is based on a hash of the entire path, to ensure uniqueness.
   *
   * You can supply custom name allocation logic by overriding the
   * `chart.generateObjectName` method.
   *
   * If you use an explicit name here, bear in mind that this reduces the
   * composability of your construct because it won't be possible to include
   * more than one instance in any app. Therefore it is highly recommended to
   * leave this unspecified.
   *
   * @default - an app-unique name generated by the chart
   */
  readonly name?: string;

  /**
   * Arbitrary key/value metadata.
   */
  readonly [key: string]: any;
}

export enum TerraformSchemaType {
  PROVIDER = 'provider',
  RESOURCE = 'resource',
  DATA = 'data'
}

export interface TerraformMetadata {
  readonly name: string;
  readonly schemaType: TerraformSchemaType;
  readonly providerVersion: string;
  readonly providerName: string;
  readonly rawSchema: string;
}

export interface IReferencableResource {
  logicalId(): string;
}
/**
 * Options for defining API objects.
 */
export interface ResourceObjectProps {
  /**
   * Data associated with the resource.
   */
  readonly data?: any;

  /**
   * Object metadata.
   *
   * If `name` is not specified, an app-unique name will be allocated by the
   * framework based on the path of the construct within thes construct tree.
   */
  readonly metadata?: ResourceObjectMetadata;  

  /**
   * Additional attributes for this API object.
   */
  readonly [key: string]: any;
}

interface TerraformMeta {
  readonly terraform: TerraformMetadata;
}

export class ResourceObject extends Construct implements TerraformMeta, IReferencableResource {  
  /**
   * The app-unique name of the object.
   *
   * The name is allocated based on the path of the object construct within the
   * construct tree.
   * 
   * @internal
   */
  public readonly _name: string;

  /**
   * Defines an API object.
   * 
   * @param scope the construct scope
   * @param ns namespace
   * @param props options
   */
  constructor(scope: Construct, ns: string, private readonly props: ResourceObjectProps, readonly terraform: TerraformMetadata) {
    super(scope, ns);
    this._name = ns;
  }

  public logicalId(): string {
    if (this.terraform.schemaType === TerraformSchemaType.DATA) {
      return `data.${snakeCase(this.terraform.name)}.${snakeCase(this._name)}`
    } else {
      return `${snakeCase(this.terraform.name)}.${snakeCase(this._name)}`
    }
  }

  protected get tfProperties(): { [key: string]: any } {
    const props = this.props || {};
    if (TagManager.isTaggable(this)) {
      const tagsProp: { [key: string]: any } = {};
      tagsProp[this.tags.tagPropertyName] = this.tags.renderTags();
      return deepMerge(props, tagsProp);
    }
    return props;
  }

  /**
   * Renders the object to Terraform config.
   * @internal
   */
  public _render(): any {
    const obj: {[k: string]: any} = {};
    obj[this._name] = this.tfProperties
    return obj
  }
}

/**
 * Merges `source` into `target`, overriding any existing values.
 * `null`s will cause a value to be deleted.
 */
function deepMerge(target: any, ...sources: any[]) {
  for (const source of sources) {
    if (typeof(source) !== 'object' || typeof(target) !== 'object') {
      throw new Error(`Invalid usage. Both source (${JSON.stringify(source)}) and target (${JSON.stringify(target)}) must be objects`);
    }

    for (const key of Object.keys(source)) {
      const value = source[key];
      if (typeof(value) === 'object' && value != null && !Array.isArray(value)) {
        // if the value at the target is not an object, override it with an
        // object so we can continue the recursion
        if (typeof(target[key]) !== 'object') {
          target[key] = {};
        }

        deepMerge(target[key], value);

        // if the result of the merge is an empty object, it's because the
        // eventual value we assigned is `undefined`, and there are no
        // sibling concrete values alongside, so we can delete this tree.
        const output = target[key];
        if (typeof(output) === 'object' && Object.keys(output).length === 0) {
          delete target[key];
        }
      } else if (value === undefined) {
        delete target[key];
      } else {
        target[key] = value;
      }
    }
  }

  return target;
}


