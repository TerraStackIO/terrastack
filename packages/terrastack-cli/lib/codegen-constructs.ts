import { JSONSchema4 } from "json-schema";
import { CodeMaker } from 'codemaker';
import { TypeGenerator } from "./codegen-types";
import {  
  pascalCase,
  camelCase,
  constantCase
} from "change-case";
import * as path from "path"

/**
 * Generates a construct for an API object defined by `def`.
 */

function getType(namespace: string) {  
  switch (namespace) {
    case 'data_source_schemas': return 'data';
    case 'resource_schemas': return 'resource';
    case 'provider': return 'provider';
    default: {
      throw new Error(`unexpected Namespace ${namespace}`);
    }
  }
}

function propertyName(property: string): string {
  // `self` doesn't work in as property name in Python
  if (property === 'self') {    
    return camelCase(`${property}Property`)
  }
  // jsii can't handle `getFoo` properties, since it's incompatible with Java
  if (property.split('_')[0] === 'get') {    
    return camelCase(property.replace('get', 'fetch'))
  }
  return camelCase(property)
}

export function emitConstructForApiObject(code: CodeMaker, def: JSONSchema4, provider: string, namespace: string, version: string) {
  const objectName = def.name;
  if (!objectName) {
    throw new Error(`no object name`);
  }

  const type = getType(namespace)  
  const scope = type === 'data' ? 'data_' : '';
  const baseName = `${scope}${objectName.toLocaleLowerCase()}`;
  const filePath = path.join(provider, namespace.replace('_schemas', ''), baseName)
  const sourceFile = `${filePath}.ts`;
  const optionsStructName = `${pascalCase(baseName)}Props`;
  const typeGenerator = new TypeGenerator(type);

  emitFile();
  return filePath;

  function emitFile() {
    code.openFile(sourceFile);
    code.line(`// generated by terrastack.io`);
    code.line();

    code.line(`import * as tst from '@terrastack/core';`);
    code.line(`import * as cdk from '@aws-cdk/core';`);
    code.line();

    emitOptionsStruct();

    code.line();

    emitConstruct();
    code.line();

    typeGenerator.generate(code);

    code.closeFile(sourceFile);
  }

  function emitOptionsStruct() {
    const copy: JSONSchema4 = { ...def };
    copy.properties = copy.properties || {};
    typeGenerator.addType(optionsStructName, copy); 
  }
  
  function emitConstruct() {
    code.line('/**');
    code.line(` * ${ def?.description }`);
    code.line(` */`);
    code.openBlock(`export class ${pascalCase(baseName)} extends tst.ResourceObject`);

    emitProperties();
    
    code.line();

    emitInitializer();
  
    code.closeBlock();
  }

  function emitInitializer() {
    code.openBlock(`public constructor(scope: cdk.Construct, ns: string, options: ${optionsStructName})`);
    emitInitializerSuper();

    emitPropsAssignment();
    code.closeBlock();
  }

  function emitInitializerSuper() {
    code.open(`super(scope, ns, { ...options }, {`);
    if (type === 'provider') {
      code.line(`name: "${provider}",`);
    } else {
      code.line(`name: "${objectName.toLocaleLowerCase()}",`);
    }
    code.line(`schemaType: tst.TerraformSchemaType.${constantCase(type)},`);
    code.line(`providerVersion: '${version}',`);
    code.line(`providerName: '${provider}',`);
    code.line(`rawSchema: \`${JSON.stringify(def).replace(new RegExp('`', 'g'), '\\`')}\``);
    code.close(`});`);
  }

  function emitPropsAssignment() {
    const propType = typeGenerator.types[`${pascalCase(baseName)}Props`]
    for (const [ propName, propSpec ] of Object.entries(propType || {})) {    
      if (propSpec.referencable) {
        if ((type !== 'data' && propName === 'id') || (type === 'data') || !propSpec.assignable) {
          code.line(`this.${propertyName(propName)} = tst.TfReference.for(this, '${propertyName(propName)}', null);`);
        } else {
          code.line(`this.${propertyName(propName)} = tst.TfReference.for(this, '${propertyName(propName)}', options.${propertyName(propName)});`);
        }
      } else {
        if (propName === 'tags') {
          code.line(`this.${propertyName(propName)} = new cdk.TagManager(cdk.TagType.MAP, '${baseName}', options.tags, { tagPropertyName: 'tags' });`);
        } else {
          code.line(`this.${propertyName(propName)} = options.${propertyName(propName)};`);
        }
      }
    }
  }

  function emitProperties() {
    const type = typeGenerator.types[`${pascalCase(baseName)}Props`]
    for (const [ propName, propSpec ] of Object.entries(type || {})) {
      const propType = (propName === 'tags') ? 'cdk.TagManager' : propSpec.type
      code.line(`${propertyName(propName)}${propSpec.referencable ? '' : '?'}: ${propType};`);
      code.line();
    }
  }
}

export function findApiObjectDefinitions(schema: JSONSchema4) {
  const result: {[k: string]: {[k: string]: Array<JSONSchema4>}} = {};
  // const result = new Array<JSONSchema4>();
  const providers = Object.keys(schema.provider_schemas)

  for (const provider of providers) {
    const providerDef = schema.provider_schemas[provider]
    for (const type of Object.keys(providerDef)) {
      if (type === 'provider') {
        if (!result[provider]) result[provider] = {};
        if (!result[provider][type]) result[provider][type] = [];
        const def = Object.assign({}, providerDef[type], {name: `${provider}_${type}`})  
        result[provider][type] = [def]
      } else {
        for (const object of Object.keys(providerDef[type])) {
          const def = Object.assign({}, providerDef[type][object], {name: object})  
          if (!result[provider]) result[provider] = {};
          if (!result[provider][type]) result[provider][type] = [];
          result[provider][type].push(def)
        }
      }      
    }
  }
  return result;
}