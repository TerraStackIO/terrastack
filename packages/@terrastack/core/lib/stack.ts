import { Construct, ISynthesisSession } from '@aws-cdk/core';
import * as fs from 'fs';
import * as path from 'path';
import { ResourceObject } from './resource-object';
import { resolve } from './_tokens';
import { removeEmpty } from './_util';
import { Names } from './names';
import {  
  snakeCase,
} from "./_util";


export class Stack extends Construct {

  /**
   * Finds the stack in which a node is defined.
   * @param node a construct node
   */
  public static of(node: Construct): Stack {
    if (node instanceof Stack) {
      return node;
    }

    const parent = node.node.scope as Construct;
    if (!parent) {
      throw new Error(`cannot find a parent chart (directly or indirectly)`);
    }

    return Stack.of(parent);
  }

  /**
   * The name of the stack's YAML file as emitted into the cloud assembly
   * directory during synthesis.
   */
  public readonly manifestFile: string;

  constructor(scope: Construct, ns: string) {
    super(scope, ns);
    this.manifestFile = `${this.node.uniqueId}.tf.json`;
  }

  /**
   * Generates a app-unique name for an object given it's construct node path.
   *
   * @param resourceObject The API object to generate a name for.
   */
  public generateObjectName(resourceObject: ResourceObject) {
    return Names.toDnsLabel(resourceObject.node.path);
  }

  protected synthesize(session: ISynthesisSession) {
    const doc: {[k: string]: {[k: string]: any}} = {};

    for (const resource of this.node.findAll()) {
      if (!(resource instanceof ResourceObject)) {
        continue;
      }
      const resourceName = snakeCase(resource.terraform.name)
      const type = snakeCase(resource.terraform.schemaType)
      

      if (!doc[type]) {
        const obj: {[k: string]: any} = {};
        doc[type] = obj
      }

      if (type === 'provider') {
        const manifest = removeEmpty(resolve(this, resource._render()));
        const merged = {...doc[type], ...manifest}
        doc[type] = merged
      } else {
        if (!doc[type][resourceName]) {
          const obj: {[k: string]: any} = {};
          doc[type][resourceName] = obj
        }
        
        const manifest = removeEmpty(resolve(this, resource._render()));
        const merged = {...doc[type][resourceName], ...manifest}
        doc[type][resourceName] = merged
      }
    }
    fs.writeFileSync(path.join(session.assembly.outdir, this.manifestFile), JSON.stringify(doc, null, 2));
  }
}