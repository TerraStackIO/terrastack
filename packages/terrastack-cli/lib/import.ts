import { promises, readFileSync, writeFileSync } from 'fs';
import { emitConstructForApiObject, findApiObjectDefinitions } from './codegen-constructs';
import { CodeMaker } from 'codemaker';
import { JSONSchema4 } from 'json-schema';
import * as path from "path"

export interface Options {
  readonly apiVersion: string;
}

export async function generateAllApiObjects(outdir: string, inputFile: string) {
  const version = '2.52.0'
  const code = new CodeMaker();
  code.indentation = 2;

  const schema = await downloadSchema(inputFile);

  const topLevelObjects = findApiObjectDefinitions(schema);

  for (const provider of Object.keys(topLevelObjects)) {
    const files = []
    for (const type of Object.keys(topLevelObjects[provider])) {
      for (const o of topLevelObjects[provider][type]) {  
        files.push(emitConstructForApiObject(code, o, provider, type, version));
      }
    }
    writeIndex(code, files, provider)
  }  
  await promises.mkdir(outdir, { recursive: true });
  await code.save(outdir);  

  for (const provider of Object.keys(topLevelObjects)) {
    writePackageJson(path.join(outdir, provider), provider)
  }
}

async function downloadSchema(inputFile: string) {
  const output = readFileSync(path.join(process.cwd(), inputFile))
  return JSON.parse(output.toString()) as JSONSchema4;
}

async function writeIndex(code: CodeMaker, files: Array<string>, filePath: string) {
  const sourceFile = path.join(filePath, 'index.ts')
  code.openFile(sourceFile);
  code.line(`// generated by terrastack.io`);
  code.line();
  files.forEach(file => {
    code.line(`export * from './${file.replace(`${filePath}/`, '')}';`);
  })
  code.line();
  code.closeFile(sourceFile);
}

async function writePackageJson(workdir: string, provider: string) {
  const main = 'index'
  const pkg = {
    name: `@terrastack/${provider}-provider`,
    version: '0.0.0',
    author: "sebastian@korfmann.net",
    main: `${main}.js`,
    types: `${main}.d.ts`,
    license: 'Apache-2.0',
    scripts: {
      build: "jsii",
      watch: "jsii -w",
      package: "jsii-pacmak",
    },
    repository: { url: 'https://github.com/terrastackio/terrastack', type: 'git' },
    jsii: {
      outdir: "dist",
      targets: {
        python: {
          distName: `terrastack_${provider}_provider`,
          module: `terrastack_${provider}_provider`
        }
      }
    },    
    dependencies: {
      "@aws-cdk/core": "^1.27.0",
      "@terrastack/core": "^0.15.0",
    },
    devDependencies: {
      "@types/node": "^13.7.7",
      jsii: "^0.22.0",
      "jsii-pacmak": "^0.22.0"
    },  
    peerDependencies: {
      "@aws-cdk/core": "^1.27.0",
      "@terrastack/core": "^0.15.0",
    }

  };
  writeFileSync(path.join(workdir, 'package.json'), JSON.stringify(pkg, undefined, 2));
}