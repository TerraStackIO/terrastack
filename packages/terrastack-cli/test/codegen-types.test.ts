import { TypeGenerator } from "../lib/codegen-types";
import { promises as fs } from 'fs';
import * as fse from 'fs-extra';
import * as path from 'path';
import * as os from 'os';
import { CodeMaker } from "codemaker";
import { JSONSchema4 } from "json-schema";

jest.setTimeout(60000); // 1min

describe('structs', () => {  
  resource('has primitive types', {
    version: 0,
    block: {
      attributes: {
        string: {
          type: "string"
        },
        bool: {
          type: "bool"
        },
        number: {
          type: "number"
        },
      }
    }
  });

  resource('has primitive optional types', {
    version: 0,
    block: {
      attributes: {
        architecture: {
          type: "string",
          optional: true
        }        
      }
    }
  });

  resource('has primitive optional computed type', {
    version: 0,
    block: {
      attributes: {
        architecture: {
          type: "string",
          optional: true,
          computed: true
        }        
      }
    }
  });

  resource('has primitive computed type', {
    version: 0,
    block: {
      attributes: {
        architecture: {
          type: "string",          
          computed: true
        }        
      }
    }
  });

  resource('has an optional map of strings', {
    version: 0,
    block: {
      attributes: {
        tags: {
          type: ["map", "string"],
          optional: true
        },
        bool: {
          type: ["map", "bool"]
        },
        number: {
          type: ["map", "number"]
        },
        numberSet: {
          type: ["set", "number"]
        }
      }
    }
  });

  resource('has list of complex nested objects', {
    version: 0,
    block: {
      attributes: {
        instance_market_options: {
          type: ["list", ["object", {
            market_type: "string",
            spot_options: ["set", ["object", {
              block_duration_minutes: "number"
            }]]
          }]],
          optional: true
        }        
      }
    }
  });

  resource('has block type nested as single', {
    version: 0,
    block: {
      attributes: {
        tags: {
          type: ["map", "string"],
          optional: true
        }        
      },
      block_types: {
        s3_import: {
          nesting_mode: "single",
          block: {
            attributes: {
              create: {
                type: "string",
                optional: true
              }
            }
          }  
        }
      }
    }
  });

  resource('has empty block type nested as single', {
    version: 0,
    block: {
      attributes: {
        tags: {
          type: ["map", "string"],
          optional: true
        }        
      },
      block_types: {
        s3_import: {
          nesting_mode: "single",
          block: {
            attributes: {}
          }  
        }
      }
    }
  });

  resource('has block type nested in block type', {
    version: 0,
    block: {
      attributes: {
        tags: {
          type: ["map", "string"],
          optional: true
        }        
      },
      block_types: {
        s3_import: {
          nesting_mode: "single",
          block: {
            block_types: {
              foo: {
                nesting_mode: "single",
                block: {
                  attributes: {
                    bar: {
                      type: "string",
                      optional: true
                    }
                  }
                }  
              }
            },
            attributes: {
              create: {
                type: "string",
                optional: true
              }
            }
          }  
        }
      }
    }
  });
});

function resource(name: string, schema: JSONSchema4) {
  which('resource', name, schema)
}


function which(type: string, name: string, schema: JSONSchema4) {
  test(name, async () => {
    const gen = new TypeGenerator(type);
    gen.addType('TestType', schema);

    await withTempDir('test', async () => {
      expect(await generate(gen)).toMatchSnapshot();
    });
  });
}

async function generate(gen: TypeGenerator) {
  const code = new CodeMaker();

  const entrypoint = 'index.ts';

  code.openFile(entrypoint);
  gen.generate(code);
  code.closeFile(entrypoint)
  await code.save('.');

  const source = await fs.readFile(path.join('.', entrypoint), 'utf-8');

  return source;
}

async function withTempDir(dirname: string, closure: () => Promise<void>) {
  const prevdir = process.cwd();
  const parent = await fs.mkdtemp(path.join(os.tmpdir(), 'terrastack.'));
  const workdir = path.join(parent, dirname);
  await fse.mkdirp(workdir);
  try {
    process.chdir(workdir);
    await closure();
  } finally {
    process.chdir(prevdir);
    await fse.remove(parent);
  }
}