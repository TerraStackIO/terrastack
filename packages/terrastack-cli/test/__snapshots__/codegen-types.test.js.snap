// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`structs has an optional map of strings 1`] = `
"export interface TestType {
    readonly tags?: {[k:string]: string};

    readonly bool: {[k:string]: boolean};

    readonly number: {[k:string]: number};

    readonly numberSet: number[];

}

"
`;

exports[`structs has block type nested as single 1`] = `
"export interface TestType {
    readonly tags?: {[k:string]: string};

    readonly s3Import?: TestTypeS3ImportProps;

}

export interface TestTypeS3ImportProps {
    readonly create?: string;

}

"
`;

exports[`structs has block type nested in block type 1`] = `
"export interface TestType {
    readonly tags?: {[k:string]: string};

    readonly s3Import?: TestTypeS3ImportProps;

}

export interface TestTypeS3ImportProps {
    readonly create?: string;

    readonly foo?: TestTypeS3ImportPropsFooProps;

}

export interface TestTypeS3ImportPropsFooProps {
    readonly bar?: string;

}

"
`;

exports[`structs has empty block type nested as single 1`] = `
"export interface TestType {
    readonly tags?: {[k:string]: string};

    readonly s3Import?: any;

}

"
`;

exports[`structs has list of complex nested objects 1`] = `
"export interface TestTypeInstanceMarketOptionsProps {
    readonly marketType: string;

    readonly spotOptions: TestTypeInstanceMarketOptionsPropsSpotOptionsProps[];

}

export interface TestTypeInstanceMarketOptionsPropsSpotOptionsProps {
    readonly blockDurationMinutes: number;

}

export interface TestType {
    readonly instanceMarketOptions?: TestTypeInstanceMarketOptionsProps[];

}

"
`;

exports[`structs has primitive computed type 1`] = `
"export interface TestType {
}

"
`;

exports[`structs has primitive optional computed type 1`] = `
"export interface TestType {
    readonly architecture?: string;

}

"
`;

exports[`structs has primitive optional types 1`] = `
"export interface TestType {
    readonly architecture?: string;

}

"
`;

exports[`structs has primitive types 1`] = `
"export interface TestType {
    readonly string: string;

    readonly bool: boolean;

    readonly number: number;

}

"
`;
